File: packages/entrelares_db_contracts/lib/models/family_invitation.dart
/// F-15: a co-caregiver invitation. Mirrors
/// `Entrelares/Models/FamilyInvitation.cs`.
///
/// Rows are written EXCLUSIVELY by the `create_invitation` / `revoke_invitation`
/// RPCs — the client only reads them (RLS: family-scoped SELECT). The two
/// computed flags below are what the Família page sorts the list by.
class FamilyInvitation {
  final int id;

  /// Which family issued it. The app only ever lists its own (RLS); the gate
  /// needs the column to sweep a family's open invitations clean between
  /// scenarios, and to prove a purge did not reach across families.
  final int? familyId;

  /// When it was issued. The S-15/A-4 purge promise ("permanently erased within
  /// 30 days") is measured from HERE, so the gate backdates this column to age
  /// an invitation instead of waiting a month.
  final DateTime? createdAt;

  final String email;
  final int roleId;

  /// The uuid that travels in `/register?invite=…`.
  final String token;

  final DateTime expiresAt;
  final DateTime? acceptedAt;
  final DateTime? revokedAt;

  const FamilyInvitation({
    required this.id,
    this.familyId,
    this.createdAt,
    required this.email,
    required this.roleId,
    required this.token,
    required this.expiresAt,
    this.acceptedAt,
    this.revokedAt,
  });

  /// Still usable: nobody accepted it, nobody revoked it, and the 7-day window
  /// has not closed. A pending invitation HOLDS A SEAT (F-37).
  bool isPending(DateTime nowUtc) =>
      acceptedAt == null && revokedAt == null && expiresAt.isAfter(nowUtc);

  /// Dead by time rather than by decision — the page offers a resend instead of
  /// the link, because the link no longer works.
  bool isExpired(DateTime nowUtc) =>
      acceptedAt == null && revokedAt == null && !expiresAt.isAfter(nowUtc);

  factory FamilyInvitation.fromJson(Map<String, dynamic> json) =>
      FamilyInvitation(
        id: json['id'] as int,
        familyId: json['family_id'] as int?,
        createdAt: _utc(json['created_at'] as String?),
        email: (json['email'] as String?) ?? '',
        roleId: json['role_id'] as int,
        token: (json['token'] as String?) ?? '',
        expiresAt: DateTime.parse(json['expires_at'] as String).toUtc(),
        acceptedAt: _utc(json['accepted_at'] as String?),
        revokedAt: _utc(json['revoked_at'] as String?),
      );

  static DateTime? _utc(String? wire) =>
      wire == null ? null : DateTime.parse(wire).toUtc();
}

File: db-gate/Entrelares.DbContracts/Models/FamilyInvitation.cs
using Supabase.Postgrest.Attributes;
using Supabase.Postgrest.Models;

namespace Entrelares.Models
{
    // F-15: co-parent invitation. Rows are written exclusively by the
    // create_invitation / revoke_invitation RPCs; the client only reads
    // (RLS: family-scoped SELECT).
    [Table("family_invitations")]
    public class FamilyInvitation : BaseModel
    {
        [PrimaryKey("id", false)]
        public long Id { get; set; }

        [Column("family_id")]
        public long FamilyId { get; set; }

        [Column("email")]
        public string Email { get; set; } = string.Empty;

        [Column("role_id")]
        public long RoleId { get; set; }

        [Column("token")]
        public Guid Token { get; set; }

        [Column("invited_by")]
        public long InvitedBy { get; set; }

        [Column("created_at")]
        public DateTime CreatedAt { get; set; }

        [Column("expires_at")]
        public DateTime ExpiresAt { get; set; }

        [Column("accepted_at")]
        public DateTime? AcceptedAt { get; set; }

        [Column("revoked_at")]
        public DateTime? RevokedAt { get; set; }

        public bool IsPending => AcceptedAt is null && RevokedAt is null && ExpiresAt > DateTime.UtcNow;
        public bool IsExpired => AcceptedAt is null && RevokedAt is null && ExpiresAt <= DateTime.UtcNow;
    }
}
